tests/requestparsing/basicauth_test.go

```diff
@@ -0,0 +1,117 @@
+
+	"github.com/google/go-safeweb/testing/requesttesting"
+)
+
+type basicAuthWant struct {
+	headers  map[string][]string
+	ok       bool
+	username string
+	password string
+}
+
+var basicAuthTests = []struct {
+	name    string
+	request []byte
+	want    basicAuthWant
+}{
+	{
+		name: "Basic",
+		request: []byte("GET / HTTP/1.1\r\n" +
+			"Host: localhost:8080\r\n" +
+			"Authorization: Basic UGVsbGU6UGFzc3dvcmQ=\r\n" +
+			"\r\n"),
+		want: basicAuthWant{
+			headers:  map[string][]string{"Authorization": []string{"Basic UGVsbGU6UGFzc3dvcmQ="}},
+			ok:       true,
+			username: "Pelle",
+			password: "Password",
+		},
+	},
+	{
+		name: "NoTrailingEquals",
+		request: []byte("GET / HTTP/1.1\r\n" +
+			"Host: localhost:8080\r\n" +
+			"Authorization: Basic UGVsbGU6UGFzc3dvcmQ\r\n" +
+			"\r\n"),
+		want: basicAuthWant{
+			headers:  map[string][]string{"Authorization": []string{"Basic UGVsbGU6UGFzc3dvcmQ"}},
+			ok:       false,
+			username: "",
+			password: "",
+		},
+	},
+	{
+		name: "DoubleColon",
+		request: []byte("GET / HTTP/1.1\r\n" +
+			"Host: localhost:8080\r\n" +
+			"Authorization: Basic UGVsbGU6UGFzc3dvcmQ6UGFzc3dvcmQ=\r\n" +
+			"\r\n"),
+		want: basicAuthWant{
+			headers:  map[string][]string{"Authorization": []string{"Basic UGVsbGU6UGFzc3dvcmQ6UGFzc3dvcmQ="}},
+			ok:       true,
+			username: "Pelle",
+			password: "Password:Password",
+		},
+	},
+	{
+		name: "NotBasic",
+		request: []byte("GET / HTTP/1.1\r\n" +
+			"Host: localhost:8080\r\n" +
+			"Authorization: xasic UGVsbGU6UGFzc3dvcmQ6UGFzc3dvcmQ=\r\n" +
+			"\r\n"),
+		want: basicAuthWant{
+			headers:  map[string][]string{"Authorization": []string{"xasic UGVsbGU6UGFzc3dvcmQ6UGFzc3dvcmQ="}},
+			ok:       false,
+			username: "",
+			password: "",
+		},
+	},
+	{
+		name: "Ordering",
+		request: []byte("GET / HTTP/1.1\r\n" +
+			"Host: localhost:8080\r\n" +
+			"Authorization: basic QUFBOmFhYQ==\r\n" +
+			"Authorization: basic QkJCOmJiYg==\r\n" +
+			"\r\n"),
+		want: basicAuthWant{
+			headers:  map[string][]string{"Authorization": []string{"basic QUFBOmFhYQ==", "basic QkJCOmJiYg=="}},
+			ok:       true,
+			username: "AAA",
+			password: "aaa",
+		},
+	},
+}
+
+func TestBasicAuth(t *testing.T) {
+	for _, tt := range basicAuthTests {
+		t.Run(tt.name, func(t *testing.T) {
+			resp, err := requesttesting.MakeRequest(context.Background(), tt.request, func(r *http.Request) {
+				if diff := cmp.Diff(tt.want.headers, map[string][]string(r.Header)); diff != "" {
+					t.Errorf("r.Header mismatch (-want +got):\n%s", diff)
+				}
+
+				username, password, ok := r.BasicAuth()
+				if ok != tt.want.ok || username != tt.want.username || password != tt.want.password {
+					t.Errorf("username, password, ok := r.BasicAuth() got: %q, %q, %v want: %q, %q, %v", username, password, ok, tt.want.username, tt.want.password, tt.want.ok)
+				}
+			})
+			if err != nil {
+				t.Errorf("MakeRequest() got: %v want: nil", err)
+			}
+
+			if !bytes.HasPrefix(resp, []byte(statusOK)) {
```

Comment:
I'm on the edge with making this a test helper. I'd like to have another reviewer chime in here, so please keep this open.

Reply:
I would make it a helper that just extracts the status from the reponse and compare that string.

```
if got, want := extractStatus(resp), statusOK; got != want {
  t.Errorf(...)
}
```

Reply:
+1.